### PR TITLE
perf: Add build logic, and upgrade it

### PR DESCRIPTION
**Pull request by Jongy**
~~Based on https://github.com/Granulate/gprofiler/pull/68~~

## Description
Dockerfile stuff are based on https://github.com/Granulate/linux/blob/granulate-perf/tools/perf/Dockerfile.

@DanielShaulov note - bringing back the process comm & not thread.

## Motivation and Context
For the build logic - see #68.
For the upgrade - it's a revert of the revert https://github.com/Granulate/gprofiler/pull/14.

## How Has This Been Tested?
* [x] On <3.16 kernel - where there's no `PERF_RECORD_MISC_COMM_EXEC`, what happens
* [x] Post 3.16 - do we really get the process comm instead of thread?

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:
- [x] My code follows the code style of this project.
- [ ] I have updated the relevant documentation.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: scripts/perf_build.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+set -euo pipefail
+
+curl -sSL https://codeload.github.com/Granulate/linux/zip/40a7823cf90a7e69ce8af88d224dfdd7e371de2d -o linux.zip
```
Comment (Jongy) [inline scripts/perf_build.sh:8]: Explain in comment why downloading the zip (git is slow, this is 200MB)

**Comment by Jongy** [comment]
Rebased post https://github.com/Granulate/gprofiler/pull/68

File: scripts/perf_env.sh [hunk 2]
```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/env bash
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+set -euo pipefail
+
+apt-get update && apt-get install -y \
```
[the remark this answers is not in the document]
Comment (Jongy) [reply scripts/perf_env.sh:8]: At first step, decided not to reduce, but to match exactly what we have in Granulate/linux. See https://github.com/Granulate/gprofiler/pull/69/commits/2ae141857835b473d8d8b33a67ee76a771b0c5b2

**Comment by Jongy** [comment]
I used this snippet:
```
import ctypes
from threading import Thread


def burner():
    libc = ctypes.CDLL("libc.so.6")
    libc.prctl(15, "blabla")
    while True:
       pass


t = Thread(target=burner)
t.start()
```

On my machine:
* Testing gProfiler 1.0.4 with `--no-python`, the bottom stack ends up as `blabla`.
* Testing this build with `--no-python`, the bottom stack ends up as `python2` (I ran it with `python2 ...`)

**Comment by Jongy** [comment]
@DanielShaulov testing on CentOS 6 with these details:
```
[2021-05-16 13:32:49,008] INFO: gprofiler.utils: gProfiler Python version: 3.6.8 (default, Nov 16 2020, 16:55:22) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44)]
[2021-05-16 13:32:49,008] INFO: gprofiler.utils: gProfiler run mode: standalone_executable
[2021-05-16 13:32:49,008] INFO: gprofiler.utils: Kernel uname release: 2.6.32-754.27.1.el6.x86_64
[2021-05-16 13:32:49,008] INFO: gprofiler.utils: Kernel uname version: #1 SMP Tue Jan 28 14:11:45 UTC 2020
[2021-05-16 13:32:49,008] INFO: gprofiler.utils: Total CPUs: 1
[2021-05-16 13:32:49,009] INFO: gprofiler.utils: Total RAM: 0.97 GB
[2021-05-16 13:32:49,025] DEBUG: gprofiler.utils: Running command: (ldd --version)
[2021-05-16 13:32:49,032] DEBUG: gprofiler.utils: (['ldd', '--version']) exit code: 0
[2021-05-16 13:32:49,033] INFO: gprofiler.utils: Linux distribution: ('CentOS', '6.10', 'Final')
[2021-05-16 13:32:49,033] INFO: gprofiler.utils: libc version: ('glibc', b'2.12')
```

1.0.4 gives me the following collapsed file (running the above `burner`):
```
blabla;PyEval_EvalFrameEx 70
blabla;[/usr/lib64/libpython2.6.so.1.0] 28
blabla;PyThreadState_Swap@plt 1
flush-202:0;blk_queue_bio_[k] 1
```

and this build gives:
```
python2;[/usr/lib64/libpython2.6.so.1.0] 100
```

Perhaps `perf` fills in the missing values despite `PERF_RECORD_MISC_COMM_EXEC` missing? I ran the Python script prior to running `perf`, which means it generates those events "delayed".

Alternatively, maybe CentOS backported this perf event? I'll check

**Comment by Jongy** [comment]
Appears to be backported...
```
$ perf script --show-task-events
...
python   590 596504.478474: PERF_RECORD_COMM exec: python:590/590
....
```

The ` exec` parts tells the source.
```
	if (event->header.misc & PERF_RECORD_MISC_COMM_EXEC)
		s = " exec";
	else
		s = "";
```

I see `PERF_RECORD_MISC_COMM_EXEC` in `/usr/src/kernels/2.6.32-754.27.1.el6.x86_64/include/linux/perf_event.h` so I assume they backported it.

**Comment by Jongy** [comment]
@guybortnikov looks like the Docker cache fails for whatever reason - https://github.com/Granulate/gprofiler/pull/69/checks?check_run_id=2604047740. I'll look into it tmrw. Might make sense to make it "allow failure" tho.

**Comment by Jongy** [comment]
Tested on Ubuntu 14.04 (the same script). `3.13.0-170-generic`. I checked its `/usr` for existence of `PERF_RECORD_MISC_COMM_EXEC` - none.

This is the result, for both gProfiler 1.0.5 and this branch
```
b;[unknown];[unknown];PyEval_EvalFrameEx 100
```

(funny that `b` is not `blabla` but it does show the point)

**Comment by Jongy** [comment]
> (also signals something bad, no symbols in this case? will check)

I downgraded libunwind & elfutils back to the version we'd used earlier. Will upgrade them in another PR, not mixing stuff here.

File: scripts/perf_env.sh (hunk 2 above)
[the remark this answers is not in the document]
Comment (Jongy) [reply scripts/perf_env.sh:8]: Leaving it

**Comment by Jongy** [comment]
@guybortnikov when CRing - can double-check me and compare this Dockerfile to https://github.com/Granulate/linux/blob/granulate-perf/tools/perf/Dockerfile.

**Review by guybortnikov — APPROVED** [review]

**Pull request merged**